packages/product/src/repositories/product.ts

```diff
@@ -135,4 +142,126 @@ export class ProductRepository extends AbstractBaseRepository<Product> {
 
     return products
   }
+
+  async update(
+    data: WithRequiredProperty<ProductServiceTypes.UpdateProductDTO, "id">[],
+    context: Context = {}
+  ): Promise<Product[]> {
+    let categoryIds: string[] = []
+    let collectionIds: string[] = []
+    let typeIds: string[] = []
+
+    const manager = (context.transactionManager ??
+      this.manager_) as SqlEntityManager
+
+    data.forEach((productData) => {
+      categoryIds = categoryIds.concat(
+        productData?.categories?.map(c => c.id) || []
+      )
+
+      if (productData.collection_id) {
+        collectionIds.push(productData.collection_id)
+      }
+
+      if (productData.type_id) {
+        typeIds.push(productData.type_id)
+      }
+    })
+
+    const productsToUpdate = await manager.find(Product, {
+      id: data.map((updateData) => updateData.id)
+    })
+
+    const collectionsToAssign = collectionIds.length ? await manager.find(ProductCollection, {
+      id: collectionIds
+    }) : []
+
+    const typesToAssign = typeIds.length ? await manager.find(ProductType, {
+      id: typeIds
+    }) : []
+
+    const categoriesToAssign = categoryIds.length ? await manager.find(ProductCategory, {
+      id: categoryIds
+    }) : []
+
+    const categoriesToAssignMap = new Map<string, ProductCategory>(
+      categoriesToAssign.map((category) => [category.id, category])
+    )
+
+    const collectionsToAssignMap = new Map<string, ProductCollection>(
+      collectionsToAssign.map((collection) => [collection.id, collection])
+    )
+
+    const typesToAssignMap = new Map<string, ProductType>(
+      typesToAssign.map((type) => [type.id, type])
+    )
+
+    const productsToUpdateMap = new Map<string, Product>(
+      productsToUpdate.map((product) => [product.id, product])
+    )
+
+    const products = await Promise.all(
+      data.map(async (updateData) => {
+        const product = productsToUpdateMap.get(updateData.id)
+
+        if (!product) {
+          throw new MedusaError(
+            MedusaError.Types.NOT_FOUND,
+            `Product with id "${updateData.id}" not found`
+          )
+        }
+
+        const {
+          categories: categoriesData,
+          tags: tagsData,
+          collection_id: collectionId,
+          type_id: typeId,
+        } = updateData
+
+        delete updateData?.categories
+        delete updateData?.collection_id
+        delete updateData?.type_id
+
+        if (isDefined(categoriesData)) {
+          await product.categories.init()
+          await product.categories.removeAll()
```

Comment:
question: Is this needed every time we update a (MikroORM) Collection?

Reply:
If the goal is a replace, then yes. But now that I see it, we probably need to delete only the ones that are no longer in the existing collection, instead of recreating the the join table records everytime. I'll make the change.